### PR TITLE
Factor out bits and pieces of controller

### PR DESCRIPTION
**Pull request by hiddeco**
Goal of this PR is to factor out logical bits of the controller into packages, to increase code readability and improve testing capabilities.

### CODE REVIEW

File: controllers/helmrelease_controller.go [hunk 1]
```diff
@@ -267,14 +265,6 @@ func (r *HelmReleaseReconciler) reconcileChart(ctx context.Context, hr *v2.HelmR
 }
 
 func (r *HelmReleaseReconciler) release(ctx context.Context, log logr.Logger, hr v2.HelmRelease, source sourcev1.Source, values chartutil.Values) (v2.HelmRelease, error) {
-	// Acquire lock
```
Comment (hiddeco) [inline controllers/helmrelease_controller.go:282]: @seaneagan can you please double check if my observations are right?

Comment (seaneagan) [reply controllers/helmrelease_controller.go:282]: Yes it seems controller-runtime ensures only 1 concurrent reconcile per object and `observedGeneration` etc get updated synchronously inside each reconciliation, so makes sense to me.

**Review by seaneagan — CHANGES_REQUESTED** [review]

**Comment by hiddeco** [comment]
@seaneagan current state of the PR is an improvement, but I still have a couple of thoughts / problems I would like to share with you:

1. I think the `helmChartFromTemplate` and `helmChartRequiresUpdate` deserve their own place outside the controller, just have not found a logical place for it (yet).
2. The `handleHelmActionResult` is not really elegant, and I am wondering if we can somehow improve it. Maybe by incorporating some of the logic into the `runner`?
3. I would like to get the `reconcileRelease` method into a state where we can perform tests with it using mocks, to confirm all state determination / decision making / actions add up.
4. `composeValues` makes inefficient API calls at the moment if the `HelmRelease` contains multiple references to the same resource (but with e.g. different paths).
5. `updateStatus` seems an odd wrapper, or an odd place for the wrapper. I have however hopes that by solving 2. this will be magically resolved too :sparkles:

**Comment by seaneagan** [comment]
>     5. `updateStatus` seems an odd wrapper, or an odd place for the wrapper

The only purpose for it is to ensure `status.lastObservedTime` is set. This field provides some value, but doesn't solve all the problems it was intended to, see https://github.com/fluxcd/toolkit/discussions/217. Should we remove it for the v1beta1 release?

**Review by stefanprodan — APPROVED** [review]
LGTM

**Pull request merged**